subprojects/core/src/test/groovy/org/gradle/initialization/MixInLegacyTypesClassLoaderTest.groovy

```diff
@@ -217,15 +219,22 @@ class MixInLegacyTypesClassLoaderTest extends Specification {
 
     def compileJavaToDir(String className, String text) {
         srcDir.createDir()
-        def srcFile = srcDir.file(className + ".java")
+        def srcFile = srcDir.file(className.split(/\./)[-1] + ".java")
```

Comment:
Why did it work with `newInstance()` but not with `getConstructor().newInstance()`?

Reply:
Previously the class is package-private, so the compiled constructor is also package-private, that's why `getConstructor().newInstance()` throws an `NoSuchMethod` exception because `getConstructor()` only lists public constructors.

Changing the class to public makes the default constructor public, so `getConstructor().newInstance()` works. However, I still don't understand why `newInstance()` worked previously, I think it shouldn't work. Do you have any ideas here?

Reply:
Looks like we should use `getDeclaredConstructor` instead of `getConstructor` as stated in the Javadoc of the deprecated `newInstance()`:
```
     * <p>The call
     *
     * <pre>{@code
     * clazz.newInstance()
     * }</pre>
     *
     * can be replaced by
     *
     * <pre>{@code
     * clazz.getDeclaredConstructor().newInstance()
     * }</pre>
```

Reply:
Can you try out if this works and if yes, replace all the calls to `getConstructor()` with `getDeclaredConstructor()`?

Reply:
This would result in `IllegalAccessException`, unless `setAccessible(true)`. I'm doing some experiment and looks like `Class.newInstance()` can bypass some runtime accessibility check. Still investigating.

Reply:
Hmm... Does that mean that we need to replace the calls to `newInstance` with something that sets `setAccessible(true)` on the returned constructor in order for the code to work in the same way?